### PR TITLE
#99: Button group component

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-util-vue/issues/99 |
| Dependencies | -- |
| Decisions | • Button toggle component <br> • Button group toggle |
| Animated GIF | button-groupbutton-toggle|

### CODE REVIEW

File: vue/components/ui/molecules/button-toggle/button-toggle.vue [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        },
+        /**
+         * The value of the button.
+         */
+        value: {
+            type: Boolean,
+            default: false
+        },
+        /**
+         * The orientation of the button that modefied its borders
```
Comment (3rdvision) [inline vue/components/ui/molecules/button-toggle/button-toggle.vue:83]: typo

File: vue/components/ui/molecules/button-toggle/button-toggle.vue [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        disabled: {
+            type: Boolean,
+            default: false
+        },
+        /**
+         * Button color props.
+         */
+        buttonProps: {
+            type: Object,
+            default: () => {}
```
Comment (3rdvision) [inline vue/components/ui/molecules/button-toggle/button-toggle.vue:102]:
```suggestion
            default: () => ({})

```

File: vue/components/ui/organisms/button-group/button-group.vue [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    watch: {
+        value(value) {
+            this.valueData = value;
+        },
+        valueData(value) {
+            this.$emit("update:value", value);
+        }
+    },
+    methods: {
+        calculateOrientation(index) {
```
Comment (3rdvision) [inline vue/components/ui/organisms/button-group/button-group.vue:55]: nice one

**Review by 3rdvision — COMMENTED** [review]
https://user-images.githubusercontent.com/24736423/122436244-92332e00-cf90-11eb-9e25-94686f69e388.mov

Clean code, just needs a little improvement
I think when we toggle a button to false, it should set the value of `<button-group>` to null but it's not happening as we can see in the video above.

File: vue/components/ui/organisms/button-group/button-group.vue [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            if (index === 0) return "left";
+            if (index === this.items.length - 1) return "right";
+            return "middle";
+        },
+        onClick(event, item) {
+            this.$set(this, "valueData", item.value);
+            console.log("val", this.valueData);
+            // this.valueData = item.value;
```
Comment (3rdvision) [inline vue/components/ui/organisms/button-group/button-group.vue:63]: remove debug code

**Review by 3rdvision — APPROVED** [review]

**Review by gcandal — APPROVED** [review]

**Comment by gcandal** [comment]
@BeeMargarida conflicts

**Review by gcandal — APPROVED** [review]

**Comment by joamag** [comment]
@BeeMargarida We must handle buttons that are not select buttons (eg: regular buttons)

**Review by 3rdvision — COMMENTED** [review]
At this point it's confusing what's expected from button-toggle and button-group.

Clean code :+1: . I think there might be just a small problem when we have `toggle: false` for a button-toggle.

So I'm talking about when we set the prop `toggle: false` for a specific button-toggle of the button-group. I would expect, in that situation, for that button to behave like a button-color in which it just emits a click when it's clicked.

The problem is button-color does in fact emit the click event but then it is not passed to the parent component (button-toggle) or even the parent of the parent (button-group). See GIF below
Only button-color is emitting the event.

Peek 2021-08-04 18-31

So from my understanding we need to find a way to let the user of the component know when the user clicks a button-toggle with `toggle: false`, we have no way of knowing this if only the child of the child emits the event.

Also, please solve the changelog conficts :smile:

**Comment by veryprofessionaldodo** [comment]
@3rdvision Nice catch about the toggle. So if I understood correctly, what you're saying is that if the `button-group` has `toggle` set to `false`, then the `button-toggle` elements inside should still emit the `click` event as normal, right? If that's the case, then the latest commit addresses that, and now the event is emitted, even if the group has that variable set to `false`. 

And nice catch on the conflicts, I tend to forget to pull the target branch, I'll keep that in mind next! :D

**Comment by 3rdvision** [comment]
@veryprofessionaldodo 

I was thinking that this could be solved by not updating the button-toggle value when prop `toggle` is `false`. I was only surprised to find that this was implemented in a way that you have to control the button-toggle state value from the outside, I was expecting it to emit an `update:value` after internal valueData change along with always emitting the `click` event.


@veryprofessionaldodo After your commit, `toggle` prop is dead code in toggle-button so I think the component might need a refactor to solve this.

@BeeMargarida do you remember why state change was done in the story instead of internally in the component just like [button-toggle-icon's implementation](https://github.com/ripe-tech/ripe-components-vue/blob/master/vue/components/ui/molecules/button-icon-toggle/button-icon-toggle.vue#L90)?
```
        methods: {
            onClick(event) {
                this.valueData = !this.valueData;
            }
        },
```

I could be missing something but looks like a good approach could be doing something similar to button-icon-toggle by creating a `valueData` and emitting `update:value` on change.

**Comment by BeeMargarida** [comment]
> @veryprofessionaldodo
> 
> I was thinking that this could be solved by not updating the button-toggle value when prop `toggle` is `false`. I was only surprised to find that this was implemented in a way that you have to control the button-toggle state value from the outside, I was expecting it to emit an `update:value` after internal valueData change along with always emitting the `click` event.
> 
> @veryprofessionaldodo After your commit, `toggle` prop is dead code in button-toggle so I think the component might need a refactor to solve this.
> 
> @BeeMargarida do you remember why state change was done in the story instead of internally in the component just like [button-toggle-icon's implementation](https://github.com/ripe-tech/ripe-components-vue/blob/master/vue/components/ui/molecules/button-icon-toggle/button-icon-toggle.vue#L90)?
> 
> ```
>         methods: {
>             onClick(event) {
>                 this.valueData = !this.valueData;
>             }
>         },
> ```
> 
> I could be missing something but looks like a good approach could be doing something similar to button-icon-toggle by creating a `valueData` and emitting `update:value` on change.

Because button toggle does not save its state inside it, it is controlled by its parent, so that story had to control it.

**Comment by 3rdvision** [comment]
> Because button toggle does not save its state inside it, it is controlled by its parent, so that story had to control it. If I remember correctly

Makes sense.

@veryprofessionaldodo button-group just needs to emit the click
 
e.g. `items` prop:
```json
[
   {
      "value":"A",
      "label":"Item 1"
   },
   {
      "value":"B",
      "label":"Item 2"
   },
   {
      "value":"C",
      "label":"Item 3",
      "toggle": false
   }
]
```

When user clicks "Item 3" button, button-group should emit `click` with payload `C` and not update `valueData`.

**Comment by veryprofessionaldodo** [comment]
@3rdvision , so if I'm understanding correctly, even if the button isn't toggled, the button group should still emit a `click` with the value of the selected item?

**Comment by 3rdvision** [comment]
> @3rdvision , so if I'm understanding correctly, even if the button isn't toggled, the button group should still emit a `click` with the value of the selected item?

Yes. The only difference being that if button item has `"toggle": true` it will emit the `update:value` and the `click` while a button that is not toggled

**Review by gcandal — APPROVED** [review]

**Review by joamag — APPROVED** [review]

**Pull request merged**